source/agora/common/crypto/Schnorr.d

```diff
@@ -64,6 +64,481 @@ nothrow @nogc unittest
+    {
+        // the challenge (the Block hash, which includes the preimages)
+        auto msg = block.hashFull();
+
+        auto sig = signChallenge(key_pair.v, rand_pair.V, rand_pair.v, msg);
+
+        block.signature.R = combine(block.signature.R, sig.R);
+        block.signature.s = combine(block.signature.s, sig.s);
+
+        auto signer_index = getIndex(key_pair.V);
+        block.validators[signer_index] = true;
+    }
+
+    bool validateBlock (Block block, Block prev_block, Point[] prev_rands)
+    {
+        // the sum of all Public keys which signed the block
+        Point Ps;
+
+        // the expected R sum in the signature. Derived from the known preimages
+        // of the previous blocks, knowing the initial R from the enrollment.
+        Point exp_Rs;
+
+        foreach (idx, preimage; block.preimages)
+        {
+            // missing preimage
+            if (idx !in prev_block.preimages)
+                return false;
+
+            // set: it must match previous preimage
+            if (preimage.hashFull() != prev_block.preimages[idx])
+                return false;
+        }
+
+        foreach (idx, set; block.validators)
+        {
+            if (!set)  // validator did not sign off the block (yet)
+                continue;
+
+            // we do not have this many validators enrolled in the system
+            if (idx >= pub_keys.length)
+                return false;
+
+            // this validator did not reveal the current preimage
+            if (cast(ushort)idx !in block.preimages)
+                return false;
+
+            // this validator did not reveal the previous preimage
+            if (cast(ushort)idx !in prev_block.preimages)
+                return false;
+
+            // we do not have a known R from a previous block (R2 = R1 + X1).
+            // the previous 'if' check should have handled this already.
+            assert(idx < prev_rands.length);
+
+            // expected R based on (R2 = R1 + X1)
+            auto expected_R = prev_rands[idx]
+                + Scalar(block.preimages[cast(ushort)idx]).toPoint();
+
+            // add it to the sum of Rs
+            exp_Rs = combine(exp_Rs, expected_R);
+
+            // combine public key
+            Ps = combine(Ps, pub_keys[idx]);
+        }
+
+        // if R was manipulated then we cannot trust the signature
+        if (exp_Rs != block.signature.R)
+        {
+            writefln("Expected combined Rs: %s", exp_Rs);
+            writefln("Actual combined Rs:   %s", block.signature.R);
+            return false;
+        }
+
+        auto msg = block.hashFull();
+        return verifyChallenge(Ps, block.signature, msg);
+    }
+
+    /** enrollment data for each validator (preimages) */
+    Hash[] n1_preims = genPreimages(Scalar.random());
+    Hash[] n2_preims = genPreimages(Scalar.random());
+    Hash[] n3_preims = genPreimages(Scalar.random());
+
+    /// Block #1
+        Block block_1;
```

Comment:
Alignment off ?

Reply:
It was just to make it visually easier to distinguish, sorry.